### PR TITLE
Fix/clipboard selection race

### PR DESCRIPTION
**Pull request by Magniquick**
Hi there, me again.

## Summary

This fixes a clipboard watcher stall after many rapid clipboard updates.

The branch is split into 2 commits:

1. `fix: avoid selection read race in watch path`
2. `fix: track offer state across selection lifecycle`

## Root Cause

There were 2 problems in the long-lived watcher path.

First, after `receive()`, the code used `roundtrip()` before reading from the pipe. In a persistent watcher, that can process a newer `Selection` before the old pipe is consumed, replacing the in-flight read state. The first commit changes this to `flush()`.

Second, MIME types and offer/device state were not tracked per offer strongly enough. The second commit:
- tracks MIME types per offer
- replaces old `Selection` / `PrimarySelection` offers cleanly
- destroys and recreates the data device on `Finished`

Applied to both `ext-data-control` and `wlr-data-control`, because both paths have the same watcher structure.

## Repro

A long-lived watcher can stop receiving new clipboard contents after many rapid clipboard updates, for example repeated `wl-copy` in the same session.

I hit this in `bubcop`: https://github.com/Magniquick/bubcop

Granted, this is a bit of an edge-case scenario, but I think it is better to fix it here than debug it later in downstream apps.

## Verification

Checked with:
- `cargo test`
- `cargo test --features wlr-data-control`
- `cargo fmt --check`
- `cargo clippy --all-targets -- -D warnings`
- `cargo clippy --all-targets --features wlr-data-control -- -D warnings`

Also tested locally with repeated rapid `wl-copy` updates in one watcher session.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -463,6 +475,47 @@ impl WlClipboardListenerStream {
...
+    fn clear_offers(&mut self) {
+        if let Some(offer) = self.selection_offer.take() {
+            offer.destroy();
+        }
+        if let Some(offer) = self.primary_selection_offer.take() {
+            offer.destroy();
+        }
+
+        self.offer_mime_types.clear();
+        self.mime_types.clear();
```
[the remark this answers is not in the document]
Comment (Magniquick) [reply src/lib.rs:517]: I do not think this can happen in the current control flow.

Before this patch, the risk was the `roundtrip()` after `receive()`: once `pipereader` was set, a full roundtrip could dispatch a newer `Selection`/`Finished` and overwrite the in-flight read state before the pipe was consumed.

After this patch, once `get_clipboard_sync()` or `try_get_clipboard()` sees `pipereader.is_some()`, it no longer dispatches Wayland events in that path. It only does `flush()` and then reads the pipe. `flush()` sends pending requests but does not dispatch incoming events into the
state machine.

So `clear_offers()` can run while waiting for a pipe to be created, but not re-enter in the middle of consuming an already-created `pipereader` in this path.

Because of that, I did not clear `pipereader` / `current_type` in `clear_offers()`: doing so would be a larger behavioral change and could discard a valid in-progress transfer without a concrete repro for that case.

File: src/lib.rs [hunk 2]
```diff
@@ -752,4 +816,45 @@ impl WlClipboardListenerStreamWlr {
...
+    }
```
[the remark this answers is not in the document]
Comment (Magniquick) [reply src/lib.rs:859]: Don't think `Finished` can re-enter the watcher in the middle of consuming an already-created `pipereader` after this patch.

The problematic re-entry point was the old `roundtrip()` after `receive()`. Once `pipereader` existed, `roundtrip()` could dispatch a newer `Selection` / `Finished` and mutate the in-flight read state before the pipe was consumed.

Now that path uses `flush()` instead. In `get_clipboard_sync()` / `try_get_clipboard()`, once `pipereader.is_some()`, the code does not dispatch Wayland events anymore in that path; it only flushes outgoing requests and then reads the pipe.

Not clearing `pipereader` / `current_type` in `clear_offers()`, because that would be a larger behavior change and could drop a valid in-progress transfer without a concrete repro for that case.

File: src/dispatch.rs [hunk 3]
```diff
@@ -108,30 +112,24 @@ impl Dispatch<ext_data_control_device_v1::ExtDataControlDeviceV1, ()>
...
-                state
-                    .data_device
-                    .as_ref()
-                    .unwrap()
-                    .set_selection(Some(&source));
+                state.clear_offers();
+                if let Some(device) = state.data_device.take() {
+                    device.destroy();
+                }
+                state.set_data_device(qh);
```
Comment (Decodetalkers) [inline src/dispatch.rs:119]: why we must replace a new data device? it is not written in the comment

File: src/dispatch_wlr.rs [hunk 4]
```diff
@@ -209,7 +207,11 @@ impl Dispatch<zwlr_data_control_offer_v1::ZwlrDataControlOfferV1, ()>
         _qhandle: &wayland_client::QueueHandle<Self>,
     ) {
         if let zwlr_data_control_offer_v1::Event::Offer { mime_type } = event {
-            state.mime_types.push(mime_type);
+            state
+                .offer_mime_types
+                .entry(_proxy.id().protocol_id())
```
Comment (Decodetalkers) [inline src/dispatch_wlr.rs]: since here we use proxy, change the param name from "_proxy" to "proxy"

File: src/dispatch.rs (hunk 3 above)
Comment (Decodetalkers) [reply src/dispatch.rs:119]: I see, it is finished. it is my bad

**Review by Decodetalkers — APPROVED** [review]
Ok, it was my bad, that time, I did not understand many concept, so I remained so many bugs in this project.

Any way, thank you for your contribution. I will publish a new release later.

Happy to know there are some people using this project
